ios-base/Networking/APIService.swift

```diff
@@ -0,0 +1,180 @@
+   using the provided keyPath and return type and returning it on the onSuccess callback.
+   - Parameters:
+   - target: The TargetType used to make the request.
+   - keyPath: The keypath used to decode from JSON (if passed nil, it will try to decode from the root).
+   */
+  open func request<T>(for target: ProviderType,
+                       at keyPath: String? = nil,
+                       onSuccess: @escaping (_ result: T, _ response: Response) -> Void,
+                       onFailure: ((Error, Response?) -> Void)? = nil) where T: Codable {
+    provider.request(target, completion: { [weak self] result in
+      guard let self = self else { return }
+
+      switch result {
+      case let .success(moyaResponse):
+        self.handleSuccess(with: moyaResponse, at: keyPath, onSuccess, onFailure)
+      case let .failure(error):
+        self.handleError(with: error, onFailure)
+      }
+    })
+  }
+
+  /**
+   Makes a request to the provided target
+   */
+  open func request(for target: ProviderType,
+                    onSuccess: @escaping (_ response: Response) -> Void,
+                    onFailure: ((Error, Response?) -> Void)? = nil) {
+    provider.request(target, completion: { [weak self] result in
+      guard let self = self else { return }
+
+      switch result {
+      case let .success(moyaResponse):
+        onSuccess(moyaResponse)
+      case let .failure(error):
+        self.handleError(with: error, onFailure)
+      }
+    })
+  }
+
+  private func handleSuccess<T>(with response: Response,
+                                at keyPath: String? = nil,
+                                _ onSuccess: @escaping (_ result: T, _ response: Response) -> Void,
+                                _ onFailure: ((Error, Response?) -> Void)? = nil) where T: Decodable {
+    do {
+      let filteredResponse = try response.filterSuccessfulStatusCodes()
+      let decodedResult = try filteredResponse.map(T.self,
+                                                   atKeyPath: keyPath,
+                                                   using: jsonDecoder,
+                                                   failsOnEmptyData: true)
+
+      onSuccess(decodedResult, filteredResponse)
+    } catch (let error) {
+      handleError(with: error, onFailure)
+    }
+  }
+
+  private func handleError(with error: Error, _ onFailure: ((Error, Response?) -> Void)? = nil) {
+    guard let moyaError = error as? MoyaError else {
+      onFailure?(error, nil)
+      return
+    }
+    switch moyaError {
+    case .imageMapping(let response):
+      onFailure?(error, response)
+    case .jsonMapping(let response):
+      onFailure?(error, response)
+    case .statusCode(let response):
+      if response.statusCode == StatusCode.unauthorized {
+        AppDelegate.shared.unexpectedLogout()
+      }
+      let decodedError = APIError.from(response: response)
+      onFailure?(decodedError ?? error, response)
```

Comment:
should we still execute the failure block when the case was already handled as unauthorized ?

Reply:
It's always a good practice to not have flows that end up hanging, it would be weird if every path but the unauthorized one calls the onFailure 🤔

Reply:
Makes sense!